src/OpenIddict.Mvc/OpenIddictExtensions.cs

```diff
@@ -169,30 +195,64 @@ public static class OpenIddictExtensions {
         private class OpenIddictConvention : IControllerModelConvention {
             public void Apply(ControllerModel controller) {
                 // Ensure the convention is only applied to the intended controller.
-                Debug.Assert(controller.ControllerType != null);
-                Debug.Assert(controller.ControllerType.IsGenericType);
-                Debug.Assert(controller.ControllerType.GetGenericTypeDefinition() == typeof(OpenIddictController<,,,>));
-
-                // Note: manually updating the controller name is required
-                // to remove the ending markers added to the generic type name.
-                controller.ControllerName = "OpenIddict";
+                if (controller.ControllerType != null &&
+                    controller.ControllerType.IsGenericType &&
+                    controller.ControllerType.GetGenericTypeDefinition() == typeof(OpenIddictController<,,,>)) {
+                    // Note: manually updating the controller name is required
+                    // to remove the ending markers added to the generic type name.
+                    controller.ControllerName = "OpenIddict";
+                }
             }
         }
 
-        private class OpenIddictPart : ApplicationPart, IApplicationPartTypeProvider {
-            public OpenIddictPart(OpenIddictBuilder builder) {
-                Types = new[] {
+        private class OpenIddictPart : ApplicationPart, IApplicationPartTypeProvider, ICompilationReferencesProvider {
```

Comment:
Not a huge fan of all this sorcery, but it's pretty much the only way to work around the bug reported by Josh.